File: backend/models/complaint.py
from mongoengine import Document, StringField, FloatField, IntField, ReferenceField, BooleanField, DateTimeField
import datetime
from .user import User
from .order import Order

class Complaint(Document):
    fromUser = ReferenceField(User, required=True)
    toUser = ReferenceField(User)
    targetId = StringField()
    entityType = StringField(required=True)
    
    # is complaint or compliment
    isComplaint = BooleanField(required=True)
    message = StringField()
    
    # Star rating (0-5, 0 means no rating)
    rating = IntField(default=0, min_value=0, max_value=5)

    weight = IntField(default=1)  # vip: 2, regular: 1

    status = StringField(
        choices=["PendingReview", "Valid", "Invalid"],
        default="PendingReview"
    )

    order = ReferenceField(Order)
    created_at = DateTimeField(default=datetime.datetime.utcnow)
    resolved_at = DateTimeField()

    meta = {
        "collection": "complaints",
        "indexes": [
            "toUser",
            "fromUser",
            "status",
            "isComplaint"]
    }

    def mark_valid(self):
        self.status = "Valid"
        self.resolved_at = datetime.datetime.utcnow()
        self.resolved_at = datetime.datetime.utcnow()
        self.save()

    def mark_invalid(self):
        self.status = "Invalid"
        self.resolved_at = datetime.datetime.utcnow()
        self.save()

File: backend/models/Complaint.js
const mongoose = require('mongoose');

const complaintSchema = new mongoose.Schema({
  fromUser: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  toUser: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: false
  },
  targetId: String,
  entityType: {
    type: String,
    required: true
  },
  isComplaint: {
    type: Boolean,
    required: true
  },
  message: String,
  rating: {
    type: Number,
    default: 0,
    min: 0,
    max: 5
  },
  weight: {
    type: Number,
    default: 1
  },
  status: {
    type: String,
    enum: ["PendingReview", "Valid", "Invalid", "Escalated"],
    default: "PendingReview"
  },
  order: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Order',
    required: false
  },
  escalationNote: String,
  created_at: {
    type: Date,
    default: Date.now
  },
  resolved_at: Date
}, {
  collection: 'complaints'
});

complaintSchema.index({ toUser: 1 });
complaintSchema.index({ fromUser: 1 });
complaintSchema.index({ status: 1 });
complaintSchema.index({ isComplaint: 1 });

complaintSchema.methods.mark_valid = function() {
  this.status = "Valid";
  this.resolved_at = new Date();
  return this.save();
};

complaintSchema.methods.mark_invalid = function() {
  this.status = "Invalid";
  this.resolved_at = new Date();
  return this.save();
};

module.exports = mongoose.model('Complaint', complaintSchema);
